gtnet/sequence.py

```diff
@@ -54,57 +10,41 @@ def _get_DNA_map(vocab=None):
     chars[(basemap[ord('N')]+9)%18]
     For this to work, bases need to be ordered as they are below
     '''
-    if vocab is None:
-        vocab = ('ACYWSKDVNTGRMHB')
-    rcmap = get_revcomp_map(vocab)
     basemap = np.zeros(128, dtype=np.uint8)
-    # reverse so we store the lowest for self-complementary codes
-    for i, c in reversed(list(enumerate(vocab))):
+    for i, c in reversed(list(enumerate(chars))):  # reverse so we store the lowest for self-complementary codes
         basemap[ord(c)] = i
         basemap[ord(c.lower())] = i
     basemap[ord('x')] = basemap[ord('X')] = basemap[ord('n')]
-    return vocab, basemap, rcmap
+    return basemap
 
 
-def get_sequences(path, basemap):
-    """
-    (1) pull all the sequences from a fasta file with scikit-bio
-    (2) return a list of the sequences mapped to 8 bit integers via
-    the provided basemap -- each item in list is a numpy array
-    * this will return a jagged array *
-    """
-    seqs = [seq.values for seq in skbio.io.read(path, format='fasta')]
-    return [basemap[seq.view(np.int8)] for seq in seqs]
-
-
-def chunk_seq(seq, chunk_size, pad_value):
-    """
-    (1) determine the padding length required from provided chunk size
-    (2) pad sequence on right side only with specified padding value
-    (3) reshape the sequence to have rows with length of chunk size
-    *only padding once on full sequence -- this ensures entire sequence
-    is divisible by the chunk size*
-    """
-    padding_len = chunk_size - len(seq) % chunk_size
-    seq = np.pad(seq, (0, padding_len), constant_values=(0, pad_value))
-    return seq.reshape((-1, chunk_size))
-
+class DNA_encoder:
+    '''
+    encodes sequences
+    '''
+    def __init__(self, chars):
+        self.chars = chars
+        self.basemap = _get_DNA_map(self.chars)
 
-def get_rev_seq(seq, rcmap):
-    """
-    (1) flips the sequence and does a index lookup in one operation
-    with our reverse complement map
-    """
-    return rcmap[np.flip(seq)]
+    def encode(self, seq):
+        charar = seq.values.view(np.uint8)
+        return self.basemap[charar]
 
 
-def get_bidir_seq(fwd_seq, rcmap, chunk_size, pad_value):
-    """
-    (1) determines the reverse complement of fwd sequence
-    (2) chunk both forward and reverse sequences
-    (3) append fwd and reverse sequence into one array
-    """
-    rev_seq = get_rev_seq(fwd_seq, rcmap)
-    fwd_chunks = chunk_seq(fwd_seq, chunk_size, pad_value)
-    rev_chunks = chunk_seq(rev_seq, chunk_size, pad_value)
-    return np.append(fwd_chunks, rev_chunks, axis=0)
+def batch_sequence(sequence, window, padval, step, chars):
+    '''
+    (0) encode sequence
+    (1) determine start + end points
+    (2) create empty batches matrix of appropriate size
+    *initialize with pad value*
+    (3) populate matrix with encoded sequence information
+    '''
+    encoder = DNA_encoder(chars)
```

Comment:
Currently, the identical object gets created for every sequence we do inference on. Try to come up with a way to avoid doing this. Two options are

1. create the DNAEncoder once outside of `batch_sequence` and pass it in every time
2. make `batch_sequence` a method on DNAEncoder

I'm sure there are others, but these two came to mind.